### PR TITLE
Small Sema fixes

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/Generics/invalid.swift [hunk 1]
```diff
@@ -79,3 +79,13 @@ func takesAny(_ a: Any) {}
 func badDiagnostic3() {
   takesAny(Deli.self) // expected-error {{argument type 'Deli<_>.Type' does not conform to expected type 'Any'}}
 }
+
+// Crash with missing nested type inside concrete type
+class OuterGeneric<T> {
```
Comment (slavapestov) [inline test/Generics/invalid.swift]: @DougGregor do we have any radars for this one? :)

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by shahmishal** [comment]
@swift-ci Please smoke test

File: lib/AST/Module.cpp [hunk 2]
```diff
@@ -610,16 +610,18 @@ TypeBase::gatherAllSubstitutions(Module *module,
...
   while (parentDC->isTypeContext())
     parentDC = parentDC->getParent();
-  if (auto *outerEnv = parentDC->getGenericEnvironmentOfContext()) {
-    for (auto gp : outerEnv->getGenericParams()) {
+  if (auto *outerSig = parentDC->getGenericSignatureOfContext()) {
+    for (auto gp : outerSig->getGenericParams()) {
       auto result = substitutions.insert(
                       {gp->getCanonicalType()->castTo<GenericTypeParamType>(),
-                       outerEnv->mapTypeIntoContext(gp)});
+                       genericEnv->mapTypeIntoContext(gp)});
```
Comment (DougGregor) [inline lib/AST/Module.cpp]: Whoops, nice catch!

File: test/Generics/invalid.swift (hunk 1 above)
Comment (DougGregor) [inline test/Generics/invalid.swift]: I bet we do... somewhere!

**Pull request merged**